### ISSUE TITLE
How to support dynamic message types for top-level messages/structs?

### ISSUE THREAD

**Issue by Jarred-Sumner** [RainwayApp/bebop#65]
Bebop looks great. I'm thinking of switching to it for a game I'm building.

**Is your feature request related to a problem? Please describe.**
What's the best way to handle multiple message types (in TypeScript)? 

Say we have a number of possible messages the server might send:
```proto
message WorldUpdateMessageContainer {
  1 -> WorldMeta meta;
  2 -> Entity[] newEntities;
  3 -> EntityUpdate[] updatedEntities;
  4 -> AbstractMessage[] messages;
  5 -> uint8[] deleteEntities;
}

message AbstractMessage {
  1 -> MessageType type;
}

message UpdateMapChunkBlobMessage {
  1 -> MessageType type;
  2 -> MapChunk chunk;
}

message PlaceBlockMessage {
  1 -> MessageType type;
  2 -> UpdateBlock[] blocks;
  3 -> uint16 player;
  4 -> int16 error;
  5 -> uint16 saveCount;
}

message ChatMessage {
  1 -> MessageType type;
  2 -> string body;
  3 -> uint16 player;
  4 -> date sentAt;
  5 -> string username;
}
```

In this case, the "top-level object" would be a `WorldUpdateMessageContainer`. What's the best practice for detecting which message it sent (in `AbstractMessage`)? 

**Describe the solution you'd like**
Is this something opcodes/mirroring is for? 

**Describe alternatives you've considered**
For the current schema format, I override the function to decode the `AbstractMessage` to look at the `MessageType` value and decode the rest in the appropriate `Message`.  I could do that here too, but is there a better way? It feels hacky to edit the generated code.

It looks like this:
```ts
const maxMessageValue = Math.max(
  ...Object.keys(CompiledSchema.MessageType)
    .filter((key) => !key.endsWith("Message"))
    .map((n) => parseInt(n, 10))
);
CompiledSchema.encoderByMessageType = new Array(maxMessageValue);
CompiledSchema.decoderByMessageType = new Array(maxMessageValue);

for (var key: MessageType in CompiledSchema.MessageType) {
  if (!key.endsWith("Message")) {
    continue;
  }

  const index = parseInt(CompiledSchema.MessageType[key], 10);
  CompiledSchema.encoderByMessageType[index] = CompiledSchema[
    "encode" + key
  ].bind(CompiledSchema);
  CompiledSchema.decoderByMessageType[index] = CompiledSchema[
    "decode" + key
  ].bind(CompiledSchema);
}

CompiledSchema.decodeAbstractMessage = function (bb, _result = null) {
  var result = _result ? _result : {};

  if (!bb || !(bb instanceof CompiledSchema.ByteBuffer)) {
    bb = new CompiledSchema.ByteBuffer(bb);
  }

  while (true) {
    switch (bb.readVarUint()) {
      case 0:
        return result;

      case 1:
        result["type"] = CompiledSchema["MessageType"][bb.readVarUint()];
        return CompiledSchema.decoderByMessageType[
          CompiledSchema["MessageType"][result["type"]]
        ](bb, result);
        break;

      default:
        throw "Error!";
    }
  }
}.bind(CompiledSchema);
```

Another way to do this would be to wrap all the objects inside other object like this:
```
message AbstractMessage {
  1 ->  UpdateMapChunkBlobMessage updateMapChunkBlobMessage;
  2 -> PlaceBlockMessage placeBlockMessage;
  3 -> ChatMessage chatMessage;
  // ...
}
```

But that would get expensive to encode/decode since its many extra object allocations.

**Comment by lynn** [comment]
Hi @Jarred-Sumner! This is a good question, and not something we've settled on a solution for.

I'd go for something like your second solution, for now. I think it's not too expensive, because `message` fields are always optional, so it's perfectly fine to only provide a value for the `2 -> placeBlockMessage;` field and leave all the others blank. This would correspond to TypeScript objects with only one key like `{'placeBlockMessage': ...}` or `{'chatMessage': ...}`. (I think it might be convenient to still have a `type` field in this case.)

The problem is that this doesn't properly model what's really going on. I think an ideal solution would look something like
```
union Message {
  UpdateMapChunkBlobMessage = 1;
  PlaceBlockMessage = 2;
  ChatMessage = 3;
}
```
which defines Message as a “tagged union” or “sum type” of the types listed.

**Comment by AndrewMD5** [comment]
To add onto this one of the reasons tagged unions weren't implemented in this initial public release is that they aren't supported natively in C#. Over the weekend I plan to experiment with implementing them, but as Lynn suggested using a mono-message type shouldn't incur much overhead since null members aren't encoded.

**Comment by AndrewMD5** [comment]
quick update: internally we agreed on how to move forward with supporting tagged unions in Bebop. A feature issue will be created today, and we'll begin the process of implementing them for 3.0.0 alongside services.

**Comment by Jarred-Sumner** [comment]
Awesome!

**Comment by AndrewMD5** [comment]
Hey @Jarred-Sumner 

This will land soon in 3.0.0. You can follow the progress on #87

**Comment by AndrewMD5** [comment]
This has landed in #87 and will be released in 3.0.0.
